File: LeetCode/python/_617.MergeTwoBinaryTrees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def mergeTrees(self, t1, t2):
        """
        :type t1: TreeNode
        :type t2: TreeNode
        :rtype: TreeNode
        """
        if not t1 and not t2:
            return None
        
        elif t1 and not t2:
            return t1
        
        elif t2 and not t1:
            return t2
        
        r = TreeNode(t1.val + t2.val)
        r.left = self.mergeTrees(t1.left, t2.left)
        r.right = self.mergeTrees(t1.right, t2.right)
        return r

File: LeetCode/C_plus/_617.MergeTwoBinaryTrees.cpp
//不熟悉c++中的树结构（类值的引用）
//c++中当定义类对象是指针对象时候，就需要用到->指向类中的成员；
//当定义一般对象时候时就需要用到"."指向类中的成员。
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* mergeTrees(TreeNode* t1, TreeNode* t2) {
        if ( t1 && t2 ) {
            TreeNode* root = new TreeNode(t1->val + t2->val); 
            //声明新对象要加new
            root->left = mergeTrees(t1->left, t2->left);
            root->right = mergeTrees(t1->right, t2->right);
            return root;
        } else {
            return t1 ? t1 : t2;
        }
    }
};
